runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -40,76 +40,103 @@ const WASMER_FEATURES: wasmer::Features = wasmer::Features {
     signal_less: true,
 };
 
-pub struct Wasmer2Memory(Memory);
+#[derive(Clone)]
+pub struct Wasmer2Memory(Arc<LinearMemory>);
 
 impl Wasmer2Memory {
-    pub fn new(
-        store: &Store,
-        initial_memory_pages: u32,
-        max_memory_pages: u32,
-    ) -> Result<Self, VMError> {
-        Ok(Wasmer2Memory(
-            Memory::new(
-                store,
-                MemoryType::new(Pages(initial_memory_pages), Some(Pages(max_memory_pages)), false),
+    pub(crate) fn new(initial_memory_pages: u32, max_memory_pages: u32) -> Result<Self, VMError> {
+        let max_pages = Pages(max_memory_pages);
+        Ok(Wasmer2Memory(Arc::new(
+            LinearMemory::new(
+                &MemoryType::new(Pages(initial_memory_pages), Some(max_pages), false),
+                &MemoryStyle::Static {
+                    bound: max_pages,
+                    offset_guard_size: wasmer_types::WASM_PAGE_SIZE as u64,
+                },
             )
-            .expect("TODO creating memory cannot fail"),
-        ))
+            .expect("creating memory must not fail"),
+        )))
     }
 
-    pub fn clone(&self) -> Memory {
-        self.0.clone()
+    // Returns the pointer to memory at the specified offset and the size of the buffer starting at
+    // the returned pointer.
+    fn data_offset(&self, offset: u64) -> Option<(*mut u8, usize)> {
```

Comment:
Naively, I'd expect this to return `Option<*mut [u8]>`. Am I correct that we don't do that here because `*mut [u8]` is impossible to construct without going via `&mut [u8]`, which we rather avoid?

Reply:
I had initially made this a `&mut [u8]` but then realized this becomes a `&T -> &mut T` function which is a big red flag something is going wrong, so I went with `*mut u8` instead and left some load-bearing safety comments. Maybe in the future there's space to improve on this, but it seems… involved.